// scryfall-sdk/out/IScry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IScry = exports.SYMBOL_CARDS = exports.SYMBOL_PRINTS = exports.SYMBOL_RULINGS = exports.SYMBOL_SET = exports.SYMBOL_COST = exports.SYMBOL_TEXT = exports.RESOURCE_GENERIC_CARD_BACK = exports.ENDPOINT_FILE_3 = exports.ENDPOINT_FILE_2 = exports.ENDPOINT_FILE_1 = exports.ENDPOINT_API = void 0;
exports.ENDPOINT_API = "https://api.scryfall.com";
exports.ENDPOINT_FILE_1 = "https://cards.scryfall.io";
exports.ENDPOINT_FILE_2 = "https://c2.scryfall.com/file";
exports.ENDPOINT_FILE_3 = "https://c3.scryfall.com/file";
exports.RESOURCE_GENERIC_CARD_BACK = `${exports.ENDPOINT_FILE_2}/scryfall-errors/missing.jpg`;
exports.SYMBOL_TEXT = Symbol("TEXT");
exports.SYMBOL_COST = Symbol("COST");
exports.SYMBOL_SET = Symbol("SET");
exports.SYMBOL_RULINGS = Symbol("RULINGS");
exports.SYMBOL_PRINTS = Symbol("PRINTS");
exports.SYMBOL_CARDS = Symbol("CARDS");
var Colors;
(function (Colors) {
    Colors[Colors["W"] = 0] = "W";
    Colors[Colors["B"] = 1] = "B";
    Colors[Colors["R"] = 2] = "R";
    Colors[Colors["U"] = 3] = "U";
    Colors[Colors["G"] = 4] = "G";
})(Colors || (Colors = {}));
var IScry;
(function (IScry) {
})(IScry = exports.IScry || (exports.IScry = {}));
