// scryfall-sdk/out/api/Rulings.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Cached_1 = require("../util/Cached");
const MagicQuerier_1 = require("../util/MagicQuerier");
class Rulings extends MagicQuerier_1.default {
    byId(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.query(["cards", id, "rulings"])).data;
        });
    }
    bySet(setCode, collectorNumber) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.query(["cards", setCode, `${collectorNumber}`, "rulings"])).data;
        });
    }
    byMultiverseId(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.query(["cards/multiverse", id, "rulings"])).data;
        });
    }
    byMtgoId(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.query(["cards/mtgo", id, "rulings"])).data;
        });
    }
    byArenaId(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.query(["cards/arena", id, "rulings"])).data;
        });
    }
}
__decorate([
    Cached_1.default
], Rulings.prototype, "byId", null);
__decorate([
    Cached_1.default
], Rulings.prototype, "bySet", null);
__decorate([
    Cached_1.default
], Rulings.prototype, "byMultiverseId", null);
__decorate([
    Cached_1.default
], Rulings.prototype, "byMtgoId", null);
__decorate([
    Cached_1.default
], Rulings.prototype, "byArenaId", null);
exports.default = new Rulings;
